mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp

```diff
@@ -0,0 +1,431 @@
+                                          OpBuilder &b, Value MN, Location loc,
+                                          Value sourceOffset, Value laneId) {
+
+  // srcOffset = k_i * MN + (laneId % wmmaInputLen) + wmmaInputLen * mn_i;
+  Value inputLen = b.create<ConstantIndexOp>(loc, wmmaInsn.inputLen);
+  sourceOffset = b.create<AddIOp>(loc, sourceOffset,
+                                  b.create<RemUIOp>(loc, laneId, inputLen));
+  sourceOffset = mnb.create<AddIOp>(loc, sourceOffset,
+                                    mnb.create<MulIOp>(loc, inputLen, mn_i));
+  sourceOffset =
+      kb.create<AddIOp>(loc, sourceOffset, kb.create<MulIOp>(loc, MN, k_i));
+  return sourceOffset;
+}
+
+void WmmaEmitter::emitThreadwiseLoop(OpBuilder &b, Location loc, Value argA,
+                                     Value argB, Value bufferC,
+                                     Value regCOffset) {
+  VectorType vectorType = wmmaInsn.retType;
+  auto vectorC = b.create<memref::LoadOp>(loc, vectorType, bufferC, regCOffset);
+
+  auto mfma = b.create<amdgpu::WMMAOp>(loc, vectorType, argA, argB, vectorC);
+  auto vectorD = mfma.getDestD();
+
+  b.create<memref::StoreOp>(loc, vectorD, bufferC, regCOffset);
+}
+
+ArrayAttr WmmaEmitter::computeOutputTransforms(
+    PatternRewriter &b, Location loc, int64_t M, int64_t N, int64_t blockSize,
+    int64_t gridSize, Value regCAllocOp, Value convertedC) {
+
+  int64_t nBlocks = N / nPerBlock;
+  int64_t mBlocks = M / mPerBlock;
+  int64_t gStride = mBlocks * nBlocks;
+  int64_t nWaves = nPerBlock / nPerWave;
+
+  // High level code for this loop
+  // source: https://gpuopen.com/learn/wmma_on_rdna3/
+  //
+  // for (int ele = 0; ele < 8; ele++){
+  //    col = laneIdx % 16;
+  //    row = (2*ele + laneIdx / 16);
+  //    c[16 * row+col] = reg[ele];
+  // }
+  //
+  int64_t retNumElements = reducedVectorType.getNumElements();
+  TopDownTMBuilder splitMemoryCoords(
+      b, {"bid", "tid", "item"},
+      {gridSize, blockSize, mRepeats * nRepeats * retNumElements}, loc);
+  splitMemoryCoords.merge({"g", "m", "n"}, {0, 1, 2}, {"bid"},
+                          {gridSize / gStride, gStride / nBlocks, nBlocks});
+
+  int64_t wavesInKernelBlock = blockSize / waveSize;
+  splitMemoryCoords.merge({"wave_m", "wave_n", "m_tid", "n_tid"}, {3, 4, 5, 6},
+                          "tid",
+                          {wavesInKernelBlock / nWaves, nWaves,
+                           waveSize / wmmaInsn.inputLen, wmmaInsn.inputLen});
+
+  splitMemoryCoords.merge({"rep_i", "rep_j", "item_i"}, {7, 8, 9}, "item",
+                          {mRepeats, nRepeats, retNumElements});
+  TransformMapAttr splitMemoryCoordsAttr = splitMemoryCoords.get();
+
+  auto toMatrixC =
+      TopDownTMBuilder::below(splitMemoryCoords, splitMemoryCoordsAttr);
+
+  toMatrixC.passThrough({"gemmG"}, {0}, {"g"});
+
+  toMatrixC.embed("gemmM", 1, M, {"m", "wave_m", "m_tid", "rep_i", "item_i"},
+                  {mPerBlock, mPerWave, 1, wmmaInsn.inputLen, 2});
```

Comment:
But `m_tid` is liimited to 0 or 1 (in the case of wave32. It would be 0,1,2,3 for wave64). Basically every workitem is computing 8 values, which represents a sub-column of a 16x16 tile. For workitems 0 to 15 `m_tid` is 0 and they write to 0, 2, 4, 6, ..., 14. For workitems 16 to 31 `m_tid` is 1 and they write at positions 1,3,5,..,15 . This is outlined here: 

![image](https://user-images.githubusercontent.com/2017147/234233334-b5277268-7fcd-4ae7-adf0-122207feed20.png)

I will add a better comment to that transformation

Reply:
Ok, having looked at the diagram, this makes sense. However, given that, can we, instead of having that magic `2`, have a value called something like `itemsPerRegister` that's a function of the wave size? That'll make it easier to understand where we got that constant.

Reply:
Done, but I used `outputStride` and I set it in the `wmmaInsn`